source/common/http/path_utility.cc

```diff
@@ -52,41 +52,25 @@ bool PathUtil::canonicalPath(HeaderEntry& path_header) {
   return true;
 }
 
-/* static */
 void PathUtil::mergeSlashes(HeaderEntry& path_header) {
   const auto original_path = path_header.value().getStringView();
-  if (original_path.empty()) {
-    return;
-  }
-
-  bool has_adjacent_slashes = false;
-  for (size_t i = 1; i < original_path.size(); ++i) {
-    if (original_path[i] == '/' && original_path[i - 1] == '/') {
-      has_adjacent_slashes = true;
-      break;
-    }
-    // Only operate on path component in URL.
-    if (original_path[i] == '?') {
-      break;
-    }
-  }
-  if (!has_adjacent_slashes) {
+  // Only operate on path component in URL.
+  const size_t query_start = original_path.find('?');
+  const auto path = original_path.substr(0, query_start);
+  const auto query = absl::ClippedSubstr(original_path, query_start);
+  if (path.find("//") == absl::string_view::npos) {
     return;
   }
 
   std::string simplified_path;
   simplified_path.reserve(original_path.size());
-  for (size_t i = 0; i < original_path.size(); ++i) {
-    if (i > 0 && original_path[i] == '/' && original_path[i - 1] == '/') {
+  for (size_t i = 0; i < path.size(); ++i) {
```

Comment:
I'd still prefer using the absl functions for this to avoid having to look carefully at string-hacking code for common patterns. But absl::Substitute is actually preferred for strings known at compile-time. Just call it in a loop till there are no more occurrences of "//" (change the 'if (path.find" to a while-loop, reversing the polarity of the condition.

Reply:
I thought about this further as my 'while' suggestion would be n^2 on a string with 100 consecutive slashes.

Let's do the initial scan for "//" via if, but then we can just use split/join to remove duplicate slashes. WDYT?
```
return absl::StrCat(absl::StrJoin(absl::StrSplit(original_path, "/"), absl::SkipEmpty), "/"), query);
```

Reply:
In this case we would also need to either special-case `/` prefix or assume that all paths are always absolute.

Reply:
Yeah also trailing slash, so cases to handle are:
```
/a//b
/a//b/
a//b
a//b/
```

It's still probably simpler than the character-at-a-time scan but maybe there's a better absl util to use?

Reply:
probably good to add those cases anyway to the test-suite regardless of the algo we use

Reply:
updated, as well as added extra tests